// immutable-js/src/Immutable.js
/**
 *  Copyright (c) 2015, Jan Biasi.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 */

// Polyfills for ES5 and older Browsers
import './util/polyfill';

import { fromJS } from './util/fromJS';
import { Iterable } from './Iterable';
import { Sequence } from './Sequence';
import { Collection } from './Collection';
import { List } from './List';
import { Stack } from './Stack';
import { Map } from './Map';
import { NativeCollection } from './Native';
import { Set } from './Set';
import { Record } from './Record';
import { Range } from './Range';
import { Repeat } from './Repeat';

export default {
    Sequence: Sequence,
    Iterable: Iterable,
    Collection: Collection,
    List: List,
    Map: Map,
    Native: NativeCollection,
    Set: Set,
    Stack: Stack,
    Range: Range,
    Record: Record,
    Repeat: Repeat,
    fromJS: fromJS
}
